required_fields.node.js

```diff
@@ -19,29 +30,618 @@ function castBool(value) {
+            zone_roof.forEach((roof, index) => {
+                if(roof.roof_type === roof_type) {
+                    addErrorMessage(errorMessages[ENUMS.ERROR], `/zone/zone_roof/${index}/${type}`, conditioned_area_invalid);
+                }
+            })
+        }
+    }
+    else {
+        zone_roof.forEach((roof, index) => {
+            if(roof.roof_type === roof_type) {
+                addErrorMessage(errorMessages[ENUMS.ERROR], `/zone/zone_roof/${index}/${type}`, combined_area_invalid);
+            }
+        })
+    }
+}
+
+function checkFloorArea(zone, conditioned_footprint, errorMessages) {
+    const {zone_floor} = zone;
+    const combined_area_invalid = checkCombinedAreaInvalid(zone);
+    if(!combined_area_invalid) {
+        const combined_floor_area = getCombinedFloorArea(zone_floor);
+        const conditioned_area_invalid = checkConditionedAreaValid(combined_floor_area, conditioned_footprint, 'floor');
+        if(conditioned_area_invalid) {
+            zone_floor.forEach((floor, index) => {
+                addErrorMessage(errorMessages[ENUMS.ERROR], `/zone/zone_floor/${index}/floor_area`, conditioned_area_invalid);
+            })
+        }
+    }
+    else {
+        zone_floor.forEach((roof, index) => {
+            addErrorMessage(errorMessages[ENUMS.ERROR], `/zone/zone_floor/${index}/floor_area`, combined_area_invalid);
+        })
+    }
+}
+
+function checkFoundationLevel(zone_floor_array, errorMessages) {
+    zone_floor_array.forEach((floor, index) => {
+        const {foundation_type, foundation_insulation_level} = floor;
+        if(!nullOrUndefined.includes(foundation_type) && !nullOrUndefined.includes(foundation_insulation_level)) {
+            if(foundation_type === 'slab_on_grade' && ![0, 5].includes(foundation_insulation_level)) {
+                addErrorMessage(errorMessages[ENUMS.ERROR], `/zone/zone_floor/${index}/foundation_insulation_level`, 'Insulation must be R-0 or R-5 for Slab on Grade Foundation');
+            } else if(![0, 11, 19].includes(foundation_insulation_level)) {
+                addErrorMessage(errorMessages[ENUMS.ERROR], `/zone/zone_floor/${index}/foundation_insulation_level`, 'Insulation must be R-0, R-11, or R-19 for current foundation type');
+            }
+        }
+    });
+}
+
+function checkConditionedAreaValid(combined_area, conditioned_footprint, area_type) {
+    const min = conditioned_footprint * 0.95;
+    const max = conditioned_footprint * 2.5;
+    if(!((min < combined_area) && (combined_area < max))) {
+        return `
+            This home's minimum footprint is approximately ${conditioned_footprint}sqft, but you
+            have specified ${combined_area}sqft of total ${area_type} area. The allowed range
+            is (${Math.ceil(min)}sqft - ${Math.floor(max)}sqft).
+            Please adjust any incorrect values. *The footprint is calculated as
+            (<total area> - <conditioned basement area>) / <number of floors>
+        `;
```

Comment:
More an annoyance atm, but if it's low hanging fruit...
These multi-line messages are returning with the lines/tabs, so I get a result that's not easy to read (see below).  If this is a quick fix (put it all on one line? :P) it may be worth addressing for these cases.
```json
"/zone/zone_roof/0/ceiling_area": [
    "\n            This home's minimum footprint is approximately 250sqft, but you\n            have specified 1000sqft of total ceiling area. The allowed range\n            is (238sqft - 625sqft).\n            Please adjust any incorrect values. *The footprint is calculated as\n            (<total area> - <conditioned basement area>) / <number of floors>\n        "
],